### PR TITLE
Nuclei segmentation

### PR DESCRIPTION
**Pull request by MikeLippincott**
This PR performs nuceli segmentation using a 2.5D approach and cellpose

### CODE REVIEW

File: 1.segment_images/examples/.DS_Store
Comment (jenna-tomkinson) [inline 1.segment_images/examples/.DS_Store:1]: Recommend ignoring.

File: 1.segment_images/examples/segmentation_output/nuclei_mask_output.gif
Comment (jenna-tomkinson) [inline 1.segment_images/examples/segmentation_output/nuclei_mask_output.gif:1]: I am noticing that some of the nuclei look to develop holes. Is it holes, or is it labeled as black?

Comment (jenna-tomkinson) [reply 1.segment_images/examples/segmentation_output/nuclei_mask_output.gif:1]: Also just to confirm, this is just meant to be a binary mask, not labelling specific nuclei?

File: 1.segment_images/local_segmentation.sh
Comment (jenna-tomkinson) [inline 1.segment_images/local_segmentation.sh:1]: Very clean!

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# This notebook focuses on trying to find a way to segment cells within organoids properly.
+# The end goals is to segment cell and extract morphology features from cellprofiler.
+# These masks must be imported into cellprofiler to extract features.
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:6]: Recommend including more info on the 2.5D methodology, it is the most important part here. You might also consider moving it out into a README instead. Here is my suggestion:

```
For segmentation, we have found 3D too computationally expensive and yields less than optimal results. 
We have developed a 2.5D segmentation methodology in combination with Cellpose.
For nuclei, we take a sliding window of 3 z-slices, where we perform max-projection on these slides and perform segmentation on these slices.
We then decouple the max-projection to collect a segmentation mask per z-slice.
We then output the masks across all z-slices per image set as a stacked-TIFF file.
```

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    args = parser.parse_args()
+    window_size = args.window_size
+    clip_limit = args.clip_limit
+    input_dir = pathlib.Path(args.input_dir).resolve(strict=True)
+else:
+    print("Running in a notebook")
+    input_dir = pathlib.Path("../examples/raw_z_input/").resolve(strict=True)
+    window_size = 3
+    clip_limit = 0.05
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:65]: Recommend adding code comments here to define these variables more clearly. I know what window_size is based on being internally involved in the project but I don't know what clip_limit means.

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        help="Clip limit for the adaptive histogram equalization",
+    )
+
+    args = parser.parse_args()
+    window_size = args.window_size
+    clip_limit = args.clip_limit
+    input_dir = pathlib.Path(args.input_dir).resolve(strict=True)
+else:
+    print("Running in a notebook")
+    input_dir = pathlib.Path("../examples/raw_z_input/").resolve(strict=True)
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:63]: Recommend considering updating this to be more specific to the plate instead of the broad folder name. We will also need this for CellProfiler to extract the metadata. I have done this on my end so this can be a change later once this method has been finalized.

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+for f in files:
+    if "405" in f:
+        nuclei = io.imread(f)
+nuclei = np.array(nuclei)
+imgs = skimage.exposure.equalize_adapthist(nuclei, clip_limit=clip_limit)
+original_imgs = imgs
+print("Subsampled image shape:", imgs.shape)
+print("number of z-slices:", imgs.shape[0])
+original_z_slice_count = len(imgs)
+print("number of z slices in the original image:", original_z_slice_count)
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:93]: Recommend code comment for this loop.

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    image_stack_window = imgs[image_index : image_index + window_size]
+    if not image_stack_window.shape[0] == window_size:
+        break
+    # max project the image stack
+    image_stack_2_5D = np.append(
+        image_stack_2_5D, np.max(image_stack_window, axis=0)[np.newaxis, :, :], axis=0
+    )
+
+image_stack_2_5D = np.array(image_stack_2_5D)
+imgs = np.array(image_stack_2_5D)
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:112]: Aren't both of these lines of code doing the same thing? I am a bit confused why you need to np.array the same variable twice. Here is my suggested change, but if there is a reason for this, recommend adding code comment:
```suggestion
imgs = np.array(image_stack_2_5D)
```

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+# In[6]:
+
+
+# Load the model
+model_name = "nuclei"
+use_GPU = core.use_gpu()
+model = models.CellposeModel(gpu=use_GPU, model_type=model_name)
+
+# Perform segmentation
+# labels, details, _ = model.eval(imgs, diameter=50, channels=[0, 0])
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:127]: Recommend removing commented-out code.

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+labels, details, _ = model.eval(
+    imgs, diameter=75, channels=[0, 0], z_axis=0, stitch_threshold=0.8
+)
+
+
+# <img src="../notebook_imgs/Sliding_window_unaggregate.jpg" alt="image" width="300"/>
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:133]: Same comment here to remove commented out code unless this is doing something in a markdown cell.

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        plt.show()
+
+
+# In[9]:
+
+
+# # save the masks
+print(reconstructed_masks.shape)
+# save the masks as tiff
+mask_file_path = pathlib.Path(mask_path / "nuclei_masks.tiff").resolve()
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/0.segment_nuclei_organoids.py:190]: Is this saving the masks per image set? This looks hardcoded to save the masks as one file name. This code might just be practice code for one image set, but recommend already setting up the code to save the masks as unique names now. You can certainly say that is out-of-scope, your choice! :)

File: 1.segment_images/scripts/2.make_nuclei_segmentation_videos.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (jenna-tomkinson) [inline 1.segment_images/scripts/2.make_nuclei_segmentation_videos.py:3]: Recommend adding a title for this file/notebook. The file name says videos but is this making GIFs? Recommend being more explicit.

File: environments/cellprofiler_env.yml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  - conda-forge::openjdk
+  # these are strict because that is how it is on the CellProfiler wiki
+  # had to move up version for Jinja2 to use jupyter nbconvert
+  - conda-forge::Jinja2=3.0.3
+  - conda-forge::inflect=5.3.0
+  - conda-forge::wxpython=4.1.0
+  - conda-forge::sentry-sdk=0.18.0
+  - conda-forge::mysqlclient=1.4.4
+  - pip:
+      - cellprofiler==4.2.6
```
Comment (jenna-tomkinson) [inline environments/cellprofiler_env.yml:31]: Recommend upgrading to CellProfiler 4.2.8, developers state it will be more stable.

**Review by jenna-tomkinson — APPROVED** [review]
LGTM! Exciting to see this method! I left some suggestions and questions, but feel free to merge after addressing. 🎉

File: 1.segment_images/examples/.DS_Store
Comment (MikeLippincott) [reply 1.segment_images/examples/.DS_Store:1]: ignored now

File: 1.segment_images/examples/segmentation_output/nuclei_mask_output.gif
Comment (MikeLippincott) [reply 1.segment_images/examples/segmentation_output/nuclei_mask_output.gif:1]: Not holes, but rather a z-slice above projected from the 2.5D approach. 

These are not binary as they retain z-information still

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py (hunk 2 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/0.segment_nuclei_organoids.py:65]: I have it defined in the argparse section too

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py (hunk 3 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/0.segment_nuclei_organoids.py:63]: I have this in the real data output dir but you cant see it since it is .gitignored!

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py (hunk 4 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/0.segment_nuclei_organoids.py:93]: added

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py (hunk 6 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/0.segment_nuclei_organoids.py:126]: removed

File: 1.segment_images/scripts/0.segment_nuclei_organoids.py (hunk 8 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/0.segment_nuclei_organoids.py:189]: The file name is hardcoded, but the path is not. The path is dynamic based on the path specified in the args

File: 1.segment_images/scripts/2.make_nuclei_segmentation_videos.py (hunk 9 above)
Comment (MikeLippincott) [reply 1.segment_images/scripts/2.make_nuclei_segmentation_videos.py:3]: added

File: environments/cellprofiler_env.yml (hunk 10 above)
Comment (MikeLippincott) [reply environments/cellprofiler_env.yml:31]: opening issue. Will test in another PR
